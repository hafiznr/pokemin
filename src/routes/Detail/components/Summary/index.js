/** @jsxRuntime classic */
/** @jsx jsx */
import React from 'react';
import { 
  Image, 
  ListItem, 
  OrderedList, 
  Skeleton, 
  Table, 
  Tbody, 
  Td, 
  Tr 
} from "@chakra-ui/react"
import { jsx, css } from '@emotion/react';

import PokemonType from "../../../../components/PokemonType";

import { getPokemonNumber } from "../../../../utils";

const IMG_STYLE_PROPS = {
  mb:"8px",
  border:"1px solid #CBD5E0",
  borderRadius:"8px",
  htmlHeight:"96px",
  htmlWidth:"96px",
};

const Summary = ({loading, data}) => {
  const summaryStyles = css({
    display: 'grid',
    gridTemplateColumns: '96px auto',
    gap: '8px'
  });

  const imageWrapperStyles = css({
    display: 'flex',
    flexDirection: 'column'
  });

  
  const { 
    abilities, 
    base_experience,
    height,
    id,
    name = '', 
    sprites,
    types,
    weight 
  } = data || {};

  const tableContent = [
    {
      label: 'Number',
      value: getPokemonNumber(id)
    },
    {
      label: 'Types',
      value: types?.map((item, i) => (
        <PokemonType key={i} type={item.type.name} />
      ))
    },
    {
      label: 'Height',
      value: height
    },
    {
      label: 'Weight',
      value: weight
    },
    {
      label: 'Abilities',
      value: (
        <OrderedList textTransform="capitalize">
          {abilities?.map((item, i) => (
            <ListItem key={i}>{item.ability.name}</ListItem>
          ))}
        </OrderedList>
      )
    },
    {
      label: 'Base exp.',
      value: base_experience
    }
  ];

  const imgFallback = (
    
      <Skeleton 
        height="96px" 
        width="96px" 
        mb="8px" 
        borderRadius="8px" 
      />
  );

  return (
    <div css={summaryStyles}>
      <div css={imageWrapperStyles}>
        <Image 
          src={sprites?.front_default} 
          alt={name}
          fallback={imgFallback}
          {...IMG_STYLE_PROPS}
        />
        {sprites?.back_default && (
          <Image 
            src={sprites?.back_default} 
            alt={name} 
            fallback={imgFallback}
            {...IMG_STYLE_PROPS}
          />
        )}
      </div>
      <Table>
        <Tbody>
          {tableContent.map((content, i) => (
            <Tr key={i}>
              <Td maxWidth="84px">{content.label}</Td>
              <Td>
                {loading ? (
                  <Skeleton width="60px" height="14px"/>
                ) : (<>{content.value}</>)}</Td>
            </Tr>
          ))}
        </Tbody>
      </Table>
    </div>
  )
}

export default Summary;