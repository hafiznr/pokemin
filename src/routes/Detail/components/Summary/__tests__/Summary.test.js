import { render, screen } from '@testing-library/react';
import Summary from '../';

test('renders Summary', () => {
  const { container, rerender } = render(<Summary loading/>);

  expect(container).toMatchSnapshot();
  expect(screen.getAllByTestId('summary-skeleton')).toHaveLength(6)

  const data = {
    abilities: [
      { 
        ability: {
          name: 'blaze'
        }  
      },
      { 
        ability: {
          name: 'solar-power'
        }  
      }
    ],
    base_experience: 240,
    height: 17,
    id: 6,
    moves: [
      {
        move: {
          name: 'mega-punch'
        }
      }
    ],
    name: "charizard",
    sprites: {
      back_default: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/6.png",  
      front_default: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/front/6.png"
    },
    status: true,
    types: [
      {
        type: {
          name: 'fire'
        }
      },
      {
        type: {
          name: 'flying'
        }
      }
    ],
    weight: 905
  };

  rerender(<Summary data={data}/>)
  expect(screen.getAllByTestId('abilities')).toHaveLength(2);
});