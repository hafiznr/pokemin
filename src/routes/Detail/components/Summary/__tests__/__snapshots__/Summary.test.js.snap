// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Summary 1`] = `
<div>
  <div
    class="css-jxadsl-Summary"
  >
    <div
      class="css-5g37ib-Summary"
    >
      <div
        class="chakra-skeleton css-ibmh67"
      />
    </div>
    <table
      class="chakra-table css-0"
      role="table"
    >
      <tbody
        class="css-0"
      >
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-jhl0h9"
            role="gridcell"
          >
            Number
          </td>
          <td
            class="css-0"
            role="gridcell"
          >
            <div
              class="chakra-skeleton css-sf9g0d"
              data-testid="summary-skeleton"
            />
          </td>
        </tr>
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-jhl0h9"
            role="gridcell"
          >
            Types
          </td>
          <td
            class="css-0"
            role="gridcell"
          >
            <div
              class="chakra-skeleton css-sf9g0d"
              data-testid="summary-skeleton"
            />
          </td>
        </tr>
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-jhl0h9"
            role="gridcell"
          >
            Height
          </td>
          <td
            class="css-0"
            role="gridcell"
          >
            <div
              class="chakra-skeleton css-sf9g0d"
              data-testid="summary-skeleton"
            />
          </td>
        </tr>
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-jhl0h9"
            role="gridcell"
          >
            Weight
          </td>
          <td
            class="css-0"
            role="gridcell"
          >
            <div
              class="chakra-skeleton css-sf9g0d"
              data-testid="summary-skeleton"
            />
          </td>
        </tr>
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-jhl0h9"
            role="gridcell"
          >
            Abilities
          </td>
          <td
            class="css-0"
            role="gridcell"
          >
            <div
              class="chakra-skeleton css-sf9g0d"
              data-testid="summary-skeleton"
            />
          </td>
        </tr>
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-jhl0h9"
            role="gridcell"
          >
            Base exp.
          </td>
          <td
            class="css-0"
            role="gridcell"
          >
            <div
              class="chakra-skeleton css-sf9g0d"
              data-testid="summary-skeleton"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
